graylog2-server/src/test/java/org/graylog/plugins/views/migrations/V20200204122000_MigrateUntypedViewsToDashboardsTest.java

```diff
@@ -0,0 +1,226 @@
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.graylog.plugins.views.migrations.V20200204122000_MigrateUntypedViewsToDashboards.V20200204122000_MigrateUntypedViewsToDashboards.MigrationCompleted;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.mockito.internal.verification.VerificationModeFactory.times;
+
+public class V20200204122000_MigrateUntypedViewsToDashboardsTest {
+    private static final String COLLECTION_VIEWS = "views";
+    private static final String COLLECTION_SEARCHES = "searches";
+
+    @Rule
+    public final MongoDBInstance mongodb = MongoDBInstance.createForClass();
+    @Rule
+    public final MockitoRule mockitoRule = MockitoJUnit.rule();
+
+    @Mock
+    private ClusterConfigService clusterConfigService;
+
+    private final ObjectMapper objectMapper = new ObjectMapperProvider().get();
+
+    private Migration migration;
+    private MongoCollection<Document> viewsCollection;
+    private MongoCollection<Document> searchesCollection;
+
+    @Before
+    public void setUp() throws Exception {
+        this.viewsCollection = spy(mongodb.mongoConnection().getMongoDatabase().getCollection(COLLECTION_VIEWS));
+        this.searchesCollection = spy(mongodb.mongoConnection().getMongoDatabase().getCollection(COLLECTION_SEARCHES));
+        final MongoConnection mongoConnection = mock(MongoConnection.class);
+        when(mongoConnection.getMongoDatabase()).thenReturn(mock(MongoDatabase.class));
+        when(mongoConnection.getMongoDatabase().getCollection(COLLECTION_VIEWS)).thenReturn(viewsCollection);
+        when(mongoConnection.getMongoDatabase().getCollection(COLLECTION_SEARCHES)).thenReturn(searchesCollection);
+        this.migration = new V20200204122000_MigrateUntypedViewsToDashboards(mongoConnection, clusterConfigService);
+    }
+
+    @Test
+    public void runsIfNoViewsArePresent() {
+        this.migration.upgrade();
+    }
+
+    @Test
+    public void writesMigrationCompletedAfterSuccess() {
+        this.migration.upgrade();
+
+        final MigrationCompleted migrationCompleted = captureMigrationCompleted();
+        assertThat(migrationCompleted.viewIds()).isEmpty();
+    }
+
+    @Test
+    public void doesNotRunAgainIfMigrationHadCompletedBefore() {
+        when(clusterConfigService.get(MigrationCompleted.class)).thenReturn(MigrationCompleted.create(Collections.emptyList()));
+
+        this.migration.upgrade();
+
+        verify(clusterConfigService, never()).write(any());
```

Comment:
I think it would be even better to verify that nothing was even read. Not writing anything could result from either the migration having been completed before OR not finding any views in need of migration.

Not crucial though.

Reply:
✅